### DISCUSSION TITLE
Pull request merge method rule public preview - feedback

### DISCUSSION THREAD

**Discussion by patrick-knight** [community/community#146284] [Repositories]
You can now enforce what merge method is available on a specific branch with the new pull request merge method rule in public preview.

Choose between Merge, Squash or Rebase and ensure only selected merge methods are supported.
![GitHubScreenShot-Arc-04-12-2024-004117@2x](https://github.com/user-attachments/assets/5eff73a4-966f-4112-9423-36793625dd54)

Learn more in the [documentation](https://docs.github.com/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/available-rules-for-rulesets#require-a-pull-request-before-merging)

#### Video example of enabling the rule

https://github.com/user-attachments/assets/de98ad0e-0097-4a50-8bb0-59f9b126f39f

### Known issues
Conflicting merge methods between the repository setting and the rule will cause merges to fail.

## Feedback
We want to hear from you! Add your comments, questions, likes and dislikes below. 🔽

**Comment by DzmVasileusky** [comment]
@patrick-knight Sorry but can you share a nav path to this feature? I checked all the tabs in repository settings including General and Branches but couldn't find it.

**Reply by tdorianh** [reply]
I found it via: 
Settings -> Rules -> Rulesets -> (edit or create a branch ruleset) -> Branch rules -> Require a pull request before merging -> Show additional settings -> Allowed merge methods

**Reply by patrick-knight** [reply]
@DzmVasileusky I added a video example to the post above.

**Reply by gebeto** [reply]
@freakyblack here ya go

https://github.com/user-attachments/assets/3396f0b1-14d4-4748-9c25-3cf8c9f18524

**Comment by CrazyMeal** [comment]
This is great news 🥳 !
Do you have an idea about when it might be available through [the rulesets API](https://docs.github.com/en/rest/repos/rules?apiVersion=2022-11-28#create-a-repository-ruleset) so then in [Terraform provider](https://github.com/integrations/terraform-provider-github) ? 👀

**Reply by patrick-knight** [reply]
PR is going up soon to add it to the API docs, but the endpoint is live now. 
```
      "type": "pull_request",
      "parameters": {
        "required_approving_review_count": 0,
        "dismiss_stale_reviews_on_push": false,
        "required_reviewers": [],
        "require_code_owner_review": false,
        "require_last_push_approval": false,
        "required_review_thread_resolution": false,
        "automatic_copilot_code_review_enabled": false,
        "allowed_merge_methods": [
          "squash",
          "merge",
          "rebase"
        ]
      }
    }
  ],```

**Reply by ciaransweet** [reply]
Just came to ask the same! Any idea when the docs will get the update?

**Comment by ebinion** [comment]
Love it! Most of my initial use case was to manage staging versus production environments with Render, Netlify and the likes without having a CD. This should work perfectly.

**Comment by mathroule** [comment]
That's great news! Thanks for implementing this feature 🎉 

This works well for one aspect of our workflow, but we still encounter a failing case.

Our workflow consists of the following steps:
1. Create a `topic` branch from the `develop` branch.
2. Merge the `topic` branch into the `develop` branch using "Squash".
3. Create a `release` branch from the `develop` branch.
4. Merge the `release` branch into the `main` branch using "Merge" (to keep Git history and keep `main` in sync with `develop`).
5. Create an `apply-release-change` branch from the `main` branch (to merge back changes from `main` into `develop`).
6. Merge the `apply-release-change` branch into the `develop` branch using "Merge".

Currently, we can only set up rule to allow "Squash" and "Merge" for merging branches into `develop`, regardless if it's a `topic` or an `apply-release-change` branch.

However, there is still an opportunity for developers to merge `topic` branch using "Merge" instead of "Squash." It would be beneficial to exclude the source branch from the ruleset, allowing us to establish rules like these:
- One ruleset that permits only "Squash" for merging into `develop`, except when the source branch is `apply-release-change`.
- Another ruleset that allows only "Merge" for merging into `develop`, except when the source branch is not `apply-release-change` (i.e. when it's a `topic` branch).

**Reply by patrick-knight** [reply]
Thanks for the note @mathroule, is the "source" / "destination" branch request one that expands beyond the merge method? Are there other things you'd want to control for based on the branch being merged into?

**Reply by mathroule** [reply]
Not at the moment. We only require specific merge strategies for "target" and "source" branches. However, introducing a way to apply rulesets to specific "source/target" branch pairs would provide greater flexibility and enable new potential use cases.

**Reply by truls1502** [reply]
I just want to point out that we really need the source and target in the merge options. 

For example, everything to the develop branch by squash commit. Then, develop the branch into a stage/UAT branch by merging the commit. The same goes for stage/UAT into the main/master branch. Then, if the main/master to develop, it should be a merge commit (not squash). So +1 for strategies for "target" and source" branches.

**Reply by ffittschen** [reply]
We have a very similar scenario of creating `sync/*` branches to merge back changes done on the `release/*` branch into `main`. In general, we require squash merges into `main`, but for the case of the head ref being a `sync/*` branch, we'd like to only allow merge commits to keep git history.

In our case we'd also benefit from other settings being controlled by the head ref:
We do require the merge queue for `main`, but the merge queue is always set to squash merge. So it would be great to either disable the merge queue requirement or change the merge method if the head ref is a `sync/*` branch.

**Reply by phillipuniverse** [reply]
I’ve got the exact same case as you @truls1502, well said!

**Reply by truls1502** [reply]
@patrick-knight, do you have any idea when we will be able to test this feature with strategies for the "target" and "source" branches?

**Reply by patrick-knight** [reply]
@truls1502 We do not currently have plans at the moment for that specific functionality.

**Reply by truls1502** [reply]
Ok, please let us know when this is planned and ready to test 😊

**Reply by warmfire540** [reply]
how are you handling merge commits?  we ask devs not to MERGE w/ the button in higher environments and instead use a command to run a fast forward.  This prevents those terrible commit messages into staging/prod only

**Reply by nickhodaly-ud** [reply]
We have the same use case that would warrant more granular control based on the source/target branches. Most of our development follows a trunk-based approach, where ~90% of pull requests need to be squashed and merged into `main`. But the other ~10% of our pull requests must first be squashed and merged into a `staging` branch before they can make their way into `main`. The problem is, because our `main` branch ruleset only allows squash and merge, we are not able to keep a linear history when it comes time to merge our `staging` branch into `main` (Pull requests into the `staging` branch have already been squash + merged). If we had an option to only allow `merge commits` when a pull request is attempting to merge the `staging` branch into `main` that seems like it would solve our problem.

We are also using merge queues, which seems like it could result in additional complexity. If we wanted our `staging` -> `main` merges to happen via Pull Request + Merge Queue we would want the merge queue to use `merge commits` when the source branch is `staging` but `squash and merge` for all other source branches.

**Reply by warmfire540** [reply]
@nickhodaly-ud you would not want to squash into staging.  here's what we ended up doing which works great, even though it's working around GH limitations.

Context, we have 3 branches - development, staging, production - which match our 3 environments.   Any merges into these branches push to the environments.

1. devs make a branch - `feature/*` or `bugfix/*`
2. our policy FORCES them to squash this into development (keep it clean..)
3. so now dev has 3-4 features in it
4. we create a PR from dev into staging
  - based on our policies, the buttons are all greyed out, a user cannot click anything
  - when ready, a comment is added, or a label applied that kicks off a bot to do a fast forward into staging.
5. this is repeated for prod.

in the end of a deployment, the project branches are all linear and show that beautiful `0 | 0`

**Reply by gduliscouet-ubitransport** [reply]
I think the limitation evoked here is quite common when building a software with continous deployment.

If I understood well, with the current preview we can restrict the merge method when the **matching branch** is **either the "source" OR the "target"** of the PR ? Is that right @patrick-knight ?

For me the most interesting would be to restrict the merge method when the **matching branch** is **the "source"** of the PR

---

There are few common flows I know of, but the problems usually appear when you have more than one "long-lived" branch. And they are good reasons to have more than one:

- You start by only using a `develop` (long-lived) branch and create some feature branches (short-lived). No versioning, as soon as a feature is merged into master it'll be deployed everywhere.
- To go a bit further you can add some tags on the `develop` branch. You can then deploy older versions of the code to staging or prod, but you won't be able to hotfix prod (or staging).
- Then if you want to be able to hotfix prod, you'll start creating branches. You now have 3 long-lived branches: `develop`, `staging` and `prod`

With this workflow we have the following use cases, and the associated merge methods:

- create a feature branch from develop and merge it into develop => "Squash and merge" `my-feature` into `develop`
- decide to update staging => "Create a merge commit" `develop` into `staging`
- decide to update prod => "Create a merge commit" `staging` into `prod`
- create a hotfix branch from prod and merge it into prod =>  "Squash and merge" `my-hotfix` into `prod`
- apply this hotfix (and maybe others) by "back-merging" it into staging => "Create a merge commit" `prod` into `staging`
- same for develop => "Create a merge commit" `staging` into `develop`

We use something similar to the above, and the typical human error we regularly face are:

1. someone clicked on "Squash and merge" for merging `develop` into `staging` => this is a big problem, we would need to detect it quickly and reset the staging branch to a previous commit
2. someone clicked on "Update branch" when back-merging `prod` into `staging` => this is also a big problem as it unwantedly moves prod forward
3. someone clicked on "Create a merge commit" for merging `my-feature` into `develop` => it is just a bit annoying for the git history

So the rule we would need for the above issues are:

1. when the "source" branch is long-lived, only allow to "Create a merge commit" => must have (because we **never want to re-write the commits** of a long-lived branch)
2. when the "source" branch is long-lived, only allow to "Create a merge commit" => must have (because it only makes sens to update the "source" branch if it is a "short-lived" branch)
3. when the "source" branch is **not** long-lived, **don't** allow to "Create a merge commit" => nice to have (for a nicer git history)

PS: I don't talk about "Rebase and merge", because it is similar to "Squash and merge". In the sense that it re-writes the commits of the source branch.

**Reply by mikkelblanne** [reply]
@patrick-knight The target branch matching was a good place to start, but we really also need source branch matching. In our case, we can make do with inclusive rules (i.e. free choice from `feature/*`, only merge commit from `main`).

We're still using git flow in a few places (`develop` -> `release` -> `main`), and people are now so accustomed to squash merging when doing trunk development in other repos, that they just click merge (and the button doesn't clearly show what kind of merge was performed last), also when merging i.e. `release` -> `main`. The result in this case is worse than just an unwanted merge commit in the history: Loads of conflicts later when merging `main` -> `develop`, because the commits weren't actually recorded as merged. Then we first have to clean up by force-pushing or reverting the squash commit, none of which are great for the git history.

I realize I'm not really adding anything new here, but please, I'm rooting for source branch matching to make an appearance 🙂

**Reply by Jon889** [reply]
@patrick-knight Is there any chance of this being added?

**Reply by readyon-raphael** [reply]
Is @patrick-knight still working on this project? Maybe not but this is still a relevant piece as commented by several people. Should we all engage on a discussion as on [this one](https://github.com/orgs/community/discussions/165095)? Seems like this one just got forgotten

**Comment by Snailedlt** [comment]
I would like to see an option to set the default merge method per branch as well, as discussed in [this discussion](https://github.com/orgs/community/discussions/20583) (or more specifically [here](https://github.com/orgs/community/discussions/20583#discussioncomment-11274767)) there are times where you want to "nudge" contributors towards using a certain merging method without enforcing it. Choosing which merge method is default is one way to do that.

**Reply by patrick-knight** [reply]
Noted and thanks for continuing to advocate for that option here. I should have been more clear in my earlier response to you in the other discussion this is _a way_ we start addressing this problem space not nessicarily the only or final one.  But I want to ensure this preview does the job and then we can iterate from here.

**Reply by DzmVasileusky** [reply]
Agree, it gives flexibility.
One of the possible use cases:
Standard feature/branch approach with develop, staging, master and feature-[n]
We want `Squash` be the primary (default) method for develop for cleaner git history but sometimes we need `Merge commit`.
Imagine you have `feature-1` which wasn't merged yet and your colleague started to work on `feature-2` which needs `feature-1` code.
In this case you need to merge `feature-1` with a `Merge commit` if you don't want conflicts and noise in the `feature-2` PR

**Reply by dsasko** [reply]
I second this request. Having the freedom to choose any strategy whilst pushing a default one is what I would ultimately want to see. Are there any plans on adding this feature?

**Reply by komarovalexander** [reply]
HI @patrick-knight is there any plans to add default merge method to the branch? Just select it in the UI of PR by default, and use current strategy it if it is unavailable as a merge method in the settings. 
For our workflow in develop branch we don't need "Require a pull request before merging" rule for all cases, but by default we would love to have "squash and merge" to keep history clean in case we create a pr - sometimes developer forgot to switch back to "squash and merge" if they did "merge" in previous pr (as result history of pr's accidentally became not so clean)

**Reply by merusso** [reply]
We want to encourage users to squash by default, but we want to continue supporting merge for certain edge cases (merging changes from one long-lived branch to another, e.g. merge `release/1.0.x` into `release/2.0.x`)

**Comment by charpeni** [comment]
There's a typo in the [changelog](https://github.blog/changelog/2024-12-04-pull-request-merge-method-rule-public-preview/#:~:text=method%20rule%20is-,availible,-for%20rulesets%20at):

> The merge method rule is **availible** for rulesets at the repository, organization and the enterprise level.

`availible` -> `available`.

**Reply by patrick-knight** [reply]
🙈 Thanks just resolved that.

**Comment by edwardyoung** [comment]
@patrick-knight is it possible to bypass this setting to fall back to the merge methods allowed at the repo level?  For example, I would want developers to only `squash` (set in a rule) to the `dev` branch while allowing members of the DevOps team to have the option to use `merge` or `squash` (set at the repo level).

**Reply by patrick-knight** [reply]
Not yet, we currently do not have bypass support built-in but once that is available you should be able to craft a rule like that.

**Reply by patrick-knight** [reply]
We just added in the ability to bypass this rule and so that should now be working as expected. Please let me know one way or the other.

**Reply by edwardyoung** [reply]
Hi @patrick-knight thanks for this!  I did notice one issue when testing the bypass feature.

In my test repo settings, under Pull Requests we allow merge commits and squash merging.  

I set up a branch rule targeting our default branch where the only thing checked off is "Require a pull request before merging" and the only thing set under additional settings is "Allowed merge methods" is set to Squash.  

If I test this rule as is, I'm getting the expected results.  On a pull request, the only merge method available is Squash.  If I add myself to the bypass list, I do see both merge commits and squash merging options but if I choose merge commit, I get an error saying "The develop branch requires linear history".  

Now if I disable the rule and go back to my PR, I'm able to choose merge commit without getting the error saying "The develop branch requires linear history".  Would you know why this is happening?  I double checked my rule and "Require linear history" is not checked off.

Thanks!

**Reply by patrick-knight** [reply]
Hey @edwardyoung,
Is there the possibility a branch protection rule is also in place on the same branch setting this requirement?

**Reply by edwardyoung** [reply]
Hi @patrick-knight ,

There is a branch protection rule but "Require linear history" is not checked off.  That's the setting I would be looking for right?

**Comment by Haarolean** [comment]
Bypassing a ruleset with the merge method rule enabled doesn't work, the ruleset is still applied fully to a person included in the bypass list.
I aimed to allow everyone squash merge and rebase only for maintainers.

**Reply by patrick-knight** [reply]
That is currently a [limitation](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/available-rules-for-rulesets#require-a-pull-request-before-merging) that we are working in remedying

**Reply by patrick-knight** [reply]
We just added in the ability to bypass this rule and so that should now be working as expected. Please let me know one way or the other.

**Reply by Haarolean** [reply]
@patrick-knight this does work indeed now, thanks very much!

**Comment by sarah-gelt** [comment]
This is perfect for our team - we want to allow squash or merge in our dev branch, with only merge being allowed in staging and main. Now configured and looks great! 🙌

**Reply by heidimhurst** [reply]
Our exact same use case - we'd been wanting to enforce this for ages, but after a particularly bad mishap were happy to find this preview and give it a go! 🙏

**Reply by warmfire540** [reply]
how are you handling merge commits? we ask devs not to MERGE w/ the button in higher environments and instead use a command to run a fast forward. This prevents those terrible commit messages into staging/prod only

**Comment by travi** [comment]
I would like the ability to add conditions to the rule. 

For the use case I'm interested in, I want to enforce a merge method that does not rewrite the history of the branch being merged when it is a (pre)release branch (in the context of a library). This would help avoid mistakes when a project values maintaining provenance between versions published to the registry, the source commits they resulted in, and the versions that follow them. In [semantic-release](https://github.com/semantic-release), release channels are aligned with branches and default conventions include support for `alpha` and `beta` branches as pre-release branches. 

I would like to leave the merge method flexible for normal feature contributions, but limit the merge method when the source branch being merged into the target branch that the rule is related to is a release branch. Ideally, I would supply a list of branch names to the rule that limits the merge type and would expect that the rule could be defined even if some of the named branches do not currently exist.

**Reply by patrick-knight** [reply]
Hey @travi,

Is this similar to the asks in this thread? Where depending on the source a different behavior is applied? https://github.com/orgs/community/discussions/146284#discussioncomment-11465262

**Comment by payaaam** [comment]
We have a rule that only allows "Merge" into our main branch. We are seeing an issue where this option shows "Not enabled for this repository"? There are no other repository rules that would conflict. I was wondering if we have something incorrectly configured?

<img width="432" alt="Screenshot 2025-01-09 at 2 01 56 PM" src="https://github.com/user-attachments/assets/a571d31d-c56f-45cd-808a-cba871cb29f8" />
<img width="915" alt="Screenshot 2025-01-09 at 1 53 34 PM" src="https://github.com/user-attachments/assets/487fce9b-6cd5-413a-975a-471b6a7dca69" />

**Reply by patrick-knight** [reply]
What are the pull request merge settings enabled on the repo? That might be in conflict with the rule. 

```https://github.com/<owner>/<repo>/settings```

**Reply by payaaam** [reply]
Yep! That was it. Thank you

**Comment by gebeto** [comment]
Is it possible to configure it in the way so:
 - `development` branch can be only merged with commit to `main` branch
 - `main` branch can be only merged with commit to `development` branch
 - any other branches can be squashed to `main` or `development`
 
I need this, to be able to force merge without squashing between `main` and `development` to keep whole history tracked, and be able to do some hotfix PRs with squash ability to `main` branch (when there are already some more changes in `development` branch and we need to merge hotfix right to the `main` branch, and then merge `main` branch to `development` to deploy that hotfix also to staging environment)

**Reply by patrick-knight** [reply]
This behavior was also mentioned in the thread [here](https://github.com/orgs/community/discussions/146284#discussioncomment-11465262). 

Currently there are no plans to address that behavior as part of this preview.

**Reply by eoudyk** [reply]
Would love for this to be added in the future, sometimes when you're going fast you might "squash and merge" a release branch instead of picking the "create a merge commit" option from the dropdown and it'd be nice to have another safe guard.

**Reply by warmfire540** [reply]
With a slew of CICD, rules and guards.  We were able to work around GH.

Merged into dev only use squash, merges into higher environments the buttons are disable and they must use CD which uses fast forwarding

**Reply by a-matthew** [reply]
This is what makes this preview/feature sort of not useful in many projects. If the only difference is the choice of what's enabled for ALL branches.

This workflow presented by gebeto is a very popular GitFlow strategy. It's about habits and enforcing a structured work environment. The merge type being defaulted to the last selected option is really a bad design choice, as people may double click the same merge type (if it's allowed) across 3 entirely different branches.

One of the described 'solutions' (understandably so) in that thread is:

> based on our policies, the buttons are all greyed out, a user cannot click anything
> when ready, a comment is added, or a label applied that kicks off a bot to do a fast forward into staging

This is quite terrifying how complicated a very basic policy becomes. This definitely should be a very straightforward and simple process as it constitutes the core of GIT platform ecosystem... Please reconsider making it simple and granular in choice.

**Comment by xdannyrobertsx** [comment]
+1 on this ✨

**Comment by bfischler** [comment]
Was this removed from public preview?

**Reply by patrick-knight** [reply]
We moved it to be generally available last week.

**Comment by crfrolik** [comment]
Please allow us to choose the default, in addition to which options are allowed.

**Comment by mynameischawoo** [comment]
From which version of GitHub Enterprise Server is the allowed merge methods option available? Or is it still not supported in any version of GitHub Enterprise Server?
I checked the release notes up to version 3.17 in the following documentation, but I couldn’t find any mention of the allowed merge methods option:
https://docs.github.com/en/enterprise-server@3.17/admin/all-releases

**Comment by StellaCheong59** [comment]
**Use case:** Back-merge from release branch breaks when squash-merged

We maintain a `release/*` → `main` back-merge workflow. These PRs must use "Create a merge commit" — if someone squash-merges, Git loses the shared history between `release` and `main`, causing every subsequent back-merge PR to include already-merged diffs.

We learned this the hard way: after a few accidental squash merges, our back-merge PRs started snowballing with duplicate changes.

Current workaround: a CAUTION banner in the PR body + team convention. Fragile — one wrong click undoes it.

Request: Please add source branch pattern matching to the merge method rule. For example, allow configuring `release/*` → `main` = merge commit only, while `feature/*` → `main` = squash only. This would make the merge method rule actually usable for teams with release branch workflows.

**Reply by gduliscouet-ubitransport** [reply]
@patrick-knight I think there is really a consensus on this use case: many teams have it, and I wouldn't want to know how many accumulated hours of developers were spent fixing their project branches after a wrong click on squash/rebase and merge.

Could you tell us if this use case is taken into account by the product team? Is the idea of matching the source branch of the merge being considered?

Thank you for helping us in making GitHub an even better place :)

**Comment by jo-takao** [comment]
This update is helpful, but it still does not solve a major problem for teams.

GitHub remembers the last merge method used by each user, and this becomes the default for all future PRs. This often leads to accidental merges with the wrong method.

For example, our workflow:
- feature → dev: requires squash merge (enforced by branch settings)
- dev → stg, stg → master: generally merge commit

However, if someone uses squash even once, the next PR to stg or master will default to squash, which is unsafe.

What we really need is the ability to explicitly set a default merge method at the repository or branch level.  
This default should take precedence over the “last used merge method” remembered per user, so that the intended merge strategy is always enforced.

Adding this would provide consistent behavior for all contributors and prevent many mistakes in real-world workflows.

**Reply by Snailedlt** [reply]
Well said!
Default merge method is a great way to nudge developers towards the right merge method without totally enforcing it.

I made a feature request for it here. Please upvote and weigh in if you have anything to add, or specific usecases and such!
https://github.com/orgs/community/discussions/20583

**Comment by wafer-bw** [comment]
@patrick-knight & @willsmythe - this does not address the [original request](https://github.com/orgs/community/discussions/10809#discussion-3844868) but it was still [closed](https://github.com/orgs/community/discussions/10809#discussioncomment-14709697)?

From original request:
> We currently merge regular PRs to the main branch using "Squash and merge" by enabling "Require linear history" option, and merging "deploy" PRs to the production branch with "Create a merge commit" in order to avoid unexpected divergent.

**Reply by warmfire540** [reply]
They are too busy artificially inflating their uptime status. Years later, we're still using a label to fast-forward PRs to avoid the merge commit... zz
